### DISCUSSION TITLE
Code refactoring

### DISCUSSION THREAD

**Discussion by m1ch** [vlachoudis/bCNC#1722] [Ideas]
I did a cleanup of the code and removing all compatibility to python < 3.8. [m1ch:bCNC](https://github.com/m1ch/bCNC/tree/Cleancodev2)

Getting a good insight in the code now, I propose to do some major refactoring as well:
- Defining 3 major python packages:
  - Core ... all basic functionality like G-code generation, etc.
  - Sender (or Com) ... all functionality that has to do with the communication with the controller 
  - GUI ... ever functionality that has to do with the GUI. No direct G-code generation, or sender calls from the GUI anymore. All calls shall be done via the Core. 
 
- Re-definition of plugins:
  - Plugins shall be code provided by the bCNC community and not necessary by the bCNC maintainers. 
  - Some plugins now are core functionality in my opinion. For example flatten, drilling, etc. Move them to core. 
  - New controllers can also be plugins.
  - Also, new senders can be plugins in future ( ie. direct IP communication)

**Comment by JipeB** [comment]
Hello,
Code cleanup is a very good thing, but it's going to have to fix a major bug that prevents bCNC from working with a decent manual tool change.
Moreover, the current interface, by its lack of intuitiveness, discourages newcomers whose only benchmarks are Wikis that are often obsolete or biased. see #1775
JP
